File: app/src/main/java/apincer/android/mmate/ui/viewmodel/MediaServerViewModel.java
package apincer.android.mmate.ui.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;
import javax.inject.Inject;
import dagger.hilt.android.lifecycle.HiltViewModel;
import apincer.android.mmate.service.MediaServerManager;
import apincer.music.core.server.spi.MediaServerHub;

@HiltViewModel
public class MediaServerViewModel extends ViewModel {

    private final MediaServerManager serverManager;

    @Inject
    public MediaServerViewModel(MediaServerManager serverManager) {
        this.serverManager = serverManager;
        // The ViewModel can bind to the service when it's created.
        serverManager.doBindService();
    }

    public LiveData<MediaServerHub.ServerStatus> getServerStatus() {
        return serverManager.getServerStatus();
    }

    public String getServerLocationUrl() {
        return serverManager.getServerLocationUrl();
    }

    public void startServer() {
        serverManager.startServer();
    }

    public void stopServer() {
        serverManager.stopServer();
    }

    // The ViewModel handles unbinding when it is cleared (i.e., when the UI is permanently gone).
    @Override
    protected void onCleared() {
        super.onCleared();
        serverManager.doUnbindService();
    }

    public String getLibraryName() {
        return serverManager.getLibraryName();
    }
}

File: app/src/main/java/apincer/android/mmate/ui/viewmodel/MediaServerViewModel.kt
package apincer.android.mmate.ui.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import apincer.android.mmate.service.MediaServerManager
import apincer.music.core.server.spi.MediaServerHub
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject

@HiltViewModel
class MediaServerViewModel @Inject constructor(
    private val serverManager: MediaServerManager
) : ViewModel() {

    init {
        serverManager.doBindService()
    }

    fun getServerStatus(): LiveData<MediaServerHub.ServerStatus> = serverManager.serverStatus

    fun getServerLocationUrl(): String? = serverManager.serverLocationUrl

    fun getLibraryName(): String? = serverManager.libraryName

    fun startServer() {
        serverManager.startServer()
    }

    fun stopServer() {
        serverManager.stopServer()
    }

    fun restartServer() {
        serverManager.restartServer()
    }

    override fun onCleared() {
        super.onCleared()
        serverManager.doUnbindService()
    }
}
